### PR TITLE
Damage rework + QOL changes

### PR DESCRIPTION
**Pull request by StrawWagen**

### CODE REVIEW

File: lua/entities/emplacements_turret_base/init.lua [hunk 1]
```diff
@@ -120,8 +120,19 @@ function ENT:GetDesiredShootPos()
...
+    local randomComponent = VectorRand( -1, 1 ) * randomMul
+    local recoilComponent = self:GetRight() * recoilMul 
+    local finalForce      = ( randomComponent + recoilComponent ) * finalMul
+    
+    self:GetPhysicsObject():ApplyForceCenter( finalForce )
+end
+
 function ENT:PhysicsSimulate( phys, deltatime )
     phys:Wake()
+    if not IsValid( self.turretBase ) then return end
```
Comment (brandonsturgeon) [inline lua/entities/emplacements_turret_base/init.lua:135]: Why was this line added?

File: lua/entities/emplacements_turret_base/shared.lua [hunk 2]
```diff
@@ -6,23 +6,39 @@ ENT.angleInverse = 1
...
-                if not IsValid( self ) then return end
-                self:EmitSound( "weapons/ar2/npc_ar2_reload.wav", 70, 50 )
-            end )
-        end
+        
+        --setup variable
+        local finalizeSoundTime = 0.1
+        local setupTime = 4
+        
+
```
Comment (brandonsturgeon) [inline lua/entities/emplacements_turret_base/shared.lua:20]:
```suggestion
```

File: lua/entities/emplacements_turret_base/shared.lua [hunk 3]
```diff
@@ -6,23 +6,39 @@ ENT.angleInverse = 1
...
+        
+        
+        if self.longSpawnSetup then
+            finalizeSoundTime = 4
+            setupTime = 9
+            if not SERVER then return end
+            self:EmitSound( "weapons/ar2/ar2_reload.wav", 70, 50 ) -- this sound plays before the final sound so that the gun isn't just sitting there doing nothing
+            
+        end 
+        timer.Simple( finalizeSoundTime, function()
```
Comment (brandonsturgeon) [inline lua/entities/emplacements_turret_base/shared.lua:28]:
```suggestion
        end 

        timer.Simple( finalizeSoundTime, function()
```

File: lua/entities/emplacements_turret_base/shared.lua [hunk 4]
```diff
@@ -6,23 +6,39 @@ ENT.angleInverse = 1
...
+            if not IsValid( self ) then return end
+            if not SERVER then return end
+            self:EmitSound( "weapons/ar2/npc_ar2_reload.wav", 70, 50 )
+
```
Comment (brandonsturgeon) [inline lua/entities/emplacements_turret_base/shared.lua:32]:
```suggestion
```

File: lua/entities/emplacements_turret_base/shared.lua [hunk 5]
```diff
@@ -6,23 +6,39 @@ ENT.angleInverse = 1
...
+            if not IsValid( self ) then return end
+            if not SERVER then return end
+            self:EmitSound( "weapons/ar2/npc_ar2_reload.wav", 70, 50 )
+            
+        end )
+        
+        timer.Simple( setupTime, function() 
+            if not IsValid( self ) then return end
+            self.doneSetup = true
+
```
Comment (brandonsturgeon) [inline lua/entities/emplacements_turret_base/shared.lua:38]:
```suggestion
```

File: lua/entities/emplacements_turret_base/shared.lua [hunk 6]
```diff
@@ -58,7 +74,22 @@ function ENT:ShooterStillValid()
         shooter = self:GetDTEntity( 0 )
     end
 
-    return IsValid( shooter ) and shooter:Alive() and ( ( self:GetPos() + self.TurretModelOffset ):Distance( shooter:GetShootPos() ) <= 60 )
+    return IsValid( shooter ) and shooter:Alive() and ( ( self:GetPos() + self.TurretModelOffset ):Distance( shooter:GetShootPos() ) <= 110 )
```
Comment (brandonsturgeon) [inline lua/entities/emplacements_turret_base/shared.lua:77]: Can we make this a variable?

File: lua/entities/emplacements_turret_base/shared.lua [hunk 7]
```diff
@@ -58,7 +74,22 @@ function ENT:ShooterStillValid()
...
+end
+
+function ENT:EmplacementDisconnect() 
+    self.Firing = false
+    self:SetShooter( nil )
```
Comment (brandonsturgeon) [inline lua/entities/emplacements_turret_base/shared.lua:82]:
```suggestion
    self:SetShooter()
```

File: lua/entities/emplacements_turret_base/shared.lua [hunk 8]
```diff
@@ -58,7 +74,22 @@ function ENT:ShooterStillValid()
...
+    self:FinishShooting()
+
+end
+
+function ENT:EmplacementConnect( plr )
+    if self.Shooter then return end -- run these once
```
Comment (brandonsturgeon) [inline lua/entities/emplacements_turret_base/shared.lua:88]:
```suggestion
    if self.Shooter then return end
```

File: lua/entities/emplacements_turret_base/shared.lua [hunk 9]
```diff
@@ -67,73 +98,72 @@ function ENT:Use( plr )
...
 
-            self:NextThink( CurTime() )
+            self.Firing = self:GetShooter():KeyDown( pressKey )
+            
+        else
+            if not SERVER then return end
+            self.OffsetAng = self.turretBase:GetAngles()
+            if not self.Shooter then return end -- run this function once
+            self:EmplacementDisconnect()
+
```
Comment (brandonsturgeon) [inline lua/entities/emplacements_turret_base/shared.lua:157]:
```suggestion
```

File: lua/entities/emplacements_turret_base/shared.lua [hunk 10]
```diff
@@ -67,73 +98,72 @@ function ENT:Use( plr )
...
+            self.OffsetPos = self.turretBase:GetAngles():Up() * 1
+            
+        end
+        
+        self:EmplacementSetupCheck()
+
+        if self:ShooterStillValid() then
+            if not self.doneSetup then 
+                self.OffsetAng = self.turretBase:GetAngles() -- makes emplacement not aim when its setting up
+                --todo, replace this with slower aiming instead
```
Comment (brandonsturgeon) [inline lua/entities/emplacements_turret_base/shared.lua:131]:
```suggestion
                -- TODO: replace this with slower aiming instead
```

File: lua/entities/emplacements_turret_base/shared.lua [hunk 11]
```diff
@@ -67,73 +98,72 @@ function ENT:Use( plr )
...
-    if not IsValid( self.turretBase ) and SERVER then
+    if not IsValid( self.turretBase ) then
+        if not SERVER then return end
         SafeRemoveEntity( self )
     else
-        if IsValid( self ) then
+        if not IsValid( self ) and IsValid( self.turretBase ) then return end
+        if SERVER then
+            self.BasePos = self.turretBase:GetPos()
+            self.OffsetPos = self.turretBase:GetAngles():Up() * 1
```
Comment (brandonsturgeon) [inline lua/entities/emplacements_turret_base/shared.lua:122]:
```suggestion
            self.OffsetPos = self.turretBase:GetAngles():Up()
```

File: lua/entities/emplacements_turret_base/shared.lua [hunk 12]
```diff
@@ -67,73 +98,72 @@ function ENT:Use( plr )
...
+
```
Comment (brandonsturgeon) [inline lua/entities/emplacements_turret_base/shared.lua:123]:
```suggestion
```

File: lua/entities/emplacements_turret_base/shared.lua [hunk 13]
```diff
@@ -67,73 +98,72 @@ function ENT:Use( plr )
...
         end
     end
 end
 
 function ENT:Think()
-    if not IsValid( self.turretBase ) and SERVER then
+    if not IsValid( self.turretBase ) then
+        if not SERVER then return end
         SafeRemoveEntity( self )
     else
```
Comment (brandonsturgeon) [inline lua/entities/emplacements_turret_base/shared.lua:118]: Something like this would let you get rid of the `else` and dedent everything inside.

It would probably be best to remove the need for this logic at all.

https://wiki.facepunch.com/gmod/Entity:CallOnRemove
```suggestion
    if not IsValid( self.turretBase ) then
        if SERVER then SafeRemoveEntity( self ) end
        return nil
    else
```

**Review by brandonsturgeon — CHANGES_REQUESTED** [review]
Some nitpick style suggestions and a few others

File: lua/entities/emplacements_turret_base/init.lua (hunk 1 above)
Comment (StrawWagen) [reply lua/entities/emplacements_turret_base/init.lua:135]: kept getting an error, related to the code below this running w/o a turretBase

File: lua/entities/emplacements_turret_base/shared.lua (hunk 6 above)
Comment (StrawWagen) [reply lua/entities/emplacements_turret_base/shared.lua:74]: Will do

Comment (StrawWagen) [reply lua/entities/emplacements_turret_base/shared.lua]: Done

File: lua/entities/emplacements_turret_base/shared.lua [hunk 14]
```diff
@@ -57,8 +70,23 @@ function ENT:ShooterStillValid()
     elseif CLIENT then
         shooter = self:GetDTEntity( 0 )
     end
+    return IsValid( shooter ) and shooter:Alive() and ( ( self:GetPos() + self.TurretModelOffset ):Distance( shooter:GetShootPos() ) <= self.emplacementDisconnectRange )
```
Comment (brandonsturgeon) [inline lua/entities/emplacements_turret_base/shared.lua:73]: Splits the logic out a bit and uses a more efficient distance method.

You could save the `maxDistanceSqr` variable wherever you set `emplacementDisconnectRange` too if you wanted.

```suggestion
    local shooterIsValid = IsValid( shooter ) and shooter:Alive()
    if not shooterIsValid then return false end

    local originPos = self:GetPos() + self.TurretModelOffset
    local maxDistanceSqr = self.emplacementDisconnectRange * self.emplacementDisconnectRange
    local distanceSqr = originPos:DistToSqr( shooter:GetShootPos() )
    
    return distanceSqr <= maxDistanceSqr
```

File: lua/entities/emplacements_turret_base/shared.lua [hunk 15]
```diff
@@ -67,73 +95,70 @@ function ENT:Use( plr )
...
         end
     end
 end
 
 function ENT:Think()
-    if not IsValid( self.turretBase ) and SERVER then
-        SafeRemoveEntity( self )
+    if not IsValid( self.turretBase ) then
+        if SERVER then SafeRemoveEntity( self ) end
+        return nil
```
Comment (brandonsturgeon) [inline lua/entities/emplacements_turret_base/shared.lua:114]: I still think it's a good idea to get rid of this logic entirely, as outlined here: https://github.com/CFC-Servers/emplacements/pull/10#discussion_r702390322

When you create the `turretBase`, you can use `turretBase:DeleteOnRemove( self )` or something like that. Then you won't need to do these checks on` :Think`

ref:
https://wiki.facepunch.com/gmod/Entity:DeleteOnRemove

File: lua/entities/emplacements_turret_base/shared.lua [hunk 16]
```diff
@@ -67,73 +95,70 @@ function ENT:Use( plr )
...
     else
-        if IsValid( self ) then
+        if not IsValid( self ) and IsValid( self.turretBase ) then return end
+        if SERVER then
+            self.BasePos = self.turretBase:GetPos()
+            self.OffsetPos = self.turretBase:GetAngles():Up()
```
Comment (brandonsturgeon) [inline lua/entities/emplacements_turret_base/shared.lua:119]: Could you set these in `PhysicsSimulate` instead? Seems better to set it up where you need it. Especially because PhysicsSimulations stop when the entity is frozen (I believe)

File: lua/entities/emplacements_turret_base/shared.lua [hunk 17]
```diff
@@ -67,73 +95,70 @@ function ENT:Use( plr )
...
+        end
+        
+        self:EmplacementSetupCheck()
+
+        if self:ShooterStillValid() then
+            if not self.doneSetup then 
+                self.OffsetAng = self.turretBase:GetAngles() -- makes emplacement not aim when its setting up
+                -- TODO: replace this with slower aiming instead
+            return end
```
Comment (brandonsturgeon) [inline lua/entities/emplacements_turret_base/shared.lua:128]:
```suggestion
                return
            end 
```

File: lua/entities/emplacements_turret_base/shared.lua [hunk 18]
```diff
@@ -67,73 +95,70 @@ function ENT:Use( plr )
...
-                    self.OffsetAng = self.turretBase:GetAngles()
-                    self:SetShooter( nil )
-                    self:FinishShooting()
+                local offsetAng = ( self:GetAttachment( self.MuzzleAttachment ).Pos - self:GetDesiredShootPos() ):GetNormal()
+                local offsetDot = ( self.turretBase:GetAngles():Right() * self.angleInverse ):Dot( offsetAng )
+                if offsetDot >= self.TurretTurnMax then
+                    local offsetAngNew = offsetAng:Angle()
+                    offsetAngNew:RotateAroundAxis( offsetAngNew:Up(), self.angleRotateAroundAxis )
+                    self.OffsetAng = offsetAngNew
                 end
```
Comment (brandonsturgeon) [inline lua/entities/emplacements_turret_base/shared.lua:137]:
```suggestion
                local muzzlePos = self:GetAttachment( self.MuzzleAttachment ).Pos
                local offsetAng = ( muzzlePos - self:GetDesiredShootPos() ):GetNormal()
                local offsetDot = ( self.turretBase:GetAngles():Right() * self.angleInverse ):Dot( offsetAng )

                if offsetDot >= self.TurretTurnMax then
                    self.OffsetAng = offsetAng:Angle()
                    offsetAngNew:RotateAroundAxis( self.OffsetAng:Up(), self.angleRotateAroundAxis )
                end
```

File: lua/entities/turret_40mm_frag/init.lua [hunk 19]
```diff
@@ -74,17 +73,23 @@ function ENT:Think()
 
             return true
         end
-
+        
+        -- damage equals 400 multiplied by two thirds of this turret's firing speed
+        local baseDamage = 185
```
Comment (brandonsturgeon) [inline lua/entities/turret_40mm_frag/init.lua:78]: Maybe you could calculate that here so you don't need the comment?

File: lua/entities/turret_bullets/shared.lua [hunk 20]
```diff
@@ -32,27 +32,30 @@ function ENT:DoShot()
...
+            local bulletDamage = 400 * self.ShotInterval -- ensuring dps of 400
             self.shootPos:FireBullets( {
                 Num = 1,
                 Src = self.shootPos:GetPos() + self.shootPos:GetAngles():Forward() * 10,
                 Dir = self.shootPos:GetAngles():Forward() * 1,
                 Spread = Vector( 0.015, 0.015, 0 ),
                 Tracer = 0,
-                Force = 2,
-                Damage = 25,
+                Force = bulletDamage,
```
Comment (brandonsturgeon) [inline lua/entities/turret_bullets/shared.lua:42]: If the damage gets too high, this Force will be come pretty crazy

**Review by brandonsturgeon — CHANGES_REQUESTED** [review]
Just a few more comments. We can do another cleanup sweep after this, seems like there are some core design issues with the addon itself

File: lua/entities/emplacements_turret_base/shared.lua [hunk 21]
```diff
@@ -6,23 +6,36 @@ ENT.angleInverse = 1
...
-        end
+        
+        --setup variable
+        local finalizeSoundTime = 0.1
+        local setupTime = 4
+        
+        if self.longSpawnSetup then
+            finalizeSoundTime = 4
+            setupTime = 9
+            if not SERVER then return end
```
[the remark this answers is not in the document]
Comment (StrawWagen) [reply lua/entities/emplacements_turret_base/shared.lua:23]: Didn't throw any errors or run weirdly when I tested it. May cause an issue but it's probably fine.

File: lua/entities/emplacements_turret_base/shared.lua (hunk 16 above)
Comment (StrawWagen) [reply lua/entities/emplacements_turret_base/shared.lua:119]: Yes, good idea, emplacement bases are always owned by world and so basically never get frozen but it's good to future-proof, maybe in the next pr?

File: lua/entities/turret_40mm_frag/init.lua (hunk 19 above)
Comment (StrawWagen) [reply lua/entities/turret_40mm_frag/init.lua:78]: I tried hardcoding the damage calculations but it felt like too much, if we wanted to change it later it would've been really weird.

File: lua/entities/turret_bullets/shared.lua (hunk 20 above)
Comment (StrawWagen) [reply lua/entities/turret_bullets/shared.lua:42]: It is quite fun to shoot something and watch it fly across half of spawn, not gonna add a clamp, make a new func or anything because it would be even more complicated, will probably just replace it with a normal value next time it needs to be edited.

File: lua/entities/emplacements_turret_base/shared.lua (hunk 15 above)
Comment (StrawWagen) [reply lua/entities/emplacements_turret_base/shared.lua:121]: That's a function? good to know.
It's not a huge impact imo if this inefficient bit of code stays here for now but DeleteOnRemove would be much better.

**Comment by StrawWagen** [comment]
Realised i had uploaded a past fix to the wrong directory, that's fixed now.
Latest ver has been tested in local multiplayer, no obvious issues.
